File: Chapter 3 (Searching problems)/Uninformed Search/Python/ucs.py
import heapq

# Weighted graph adjacency matrix: -1 means no edge; positive weight otherwise.
def ucs(graph, start, goal):
    n=len(graph)
    pq=[]
    heapq.heappush(pq,(0,start))
    dist=[float('inf')]*n
    dist[start]=0
    parent=[-1]*n
    visited=[False]*n
    while pq:
        cost,u=heapq.heappop(pq)
        if visited[u]:
            continue
        if u==goal:
            path=[]
            while u!=-1:
                path.append(u)
                u=parent[u]
            return list(reversed(path)),cost
        visited[u]=True
        for v in range(n):
            w=graph[u][v]
            if w!=-1:
                ng=dist[u]+w
                if ng<dist[v]:
                    dist[v]=ng
                    parent[v]=u
                    heapq.heappush(pq,(ng,v))
    return [],float('inf')

if __name__=="__main__":
    names=['A','B','C','D']
    n=len(names)
    g=[[-1]*n for _ in range(n)]
    def add(u,v,w): g[u][v]=w
    add(0,1,1); add(0,2,4); add(1,2,2); add(1,3,5); add(2,3,1)
    path, cost = ucs(g,0,3)
    print([names[i] for i in path], cost)

File: Chapter 3 (Searching problems)/Uninformed Search/C++/ucs.cpp
#include <bits/stdc++.h>
using namespace std;

// Uniform Cost Search on adjacency matrix with non-negative weights.
// -1 indicates no edge.
pair<vector<int>, int> ucs(const vector<vector<int>> &g, int s, int t)
{
    int n = (int)g.size();
    using T = pair<int, int>; // (cost,node)
    priority_queue<T, vector<T>, greater<T>> pq;
    pq.push({0, s});
    vector<int> dist(n, INT_MAX), par(n, -1);
    dist[s] = 0;
    vector<char> vis(n, 0);
    while (!pq.empty())
    {
        auto [c, u] = pq.top();
        pq.pop();
        if (vis[u])
            continue;
        if (u == t)
        {
            vector<int> path;
            for (int x = u; x != -1; x = par[x])
                path.push_back(x);
            reverse(path.begin(), path.end());
            return {path, c};
        }
        vis[u] = 1;
        for (int v = 0; v < n; ++v)
        {
            int w = g[u][v];
            if (w != -1)
            {
                int ng = dist[u] + w;
                if (ng < dist[v])
                {
                    dist[v] = ng;
                    par[v] = u;
                    pq.push({ng, v});
                }
            }
        }
    }
    return {{}, INT_MAX};
}

int main()
{
    vector<string> names = {"A", "B", "C", "D"};
    int n = (int)names.size();
    vector<vector<int>> g(n, vector<int>(n, -1));
    auto add_edge = [&](int u, int v, int w)
    { g[u][v] = w; };
    add_edge(0, 1, 1);
    add_edge(0, 2, 4);
    add_edge(1, 2, 2);
    add_edge(1, 3, 5);
    add_edge(2, 3, 1);

    auto res = ucs(g, 0, 3);
    for (size_t i = 0; i < res.first.size(); ++i)
    {
        if (i)
            cout << " ";
        cout << names[res.first[i]];
    }
    cout << " \n"
         << res.second << "\n";
}
